presto-spark-base/src/main/java/com/facebook/presto/spark/PrestoSparkQueryExecutionFactory.java

```diff
@@ -275,18 +317,25 @@ private PrestoSparkQueryExecution(
             List<Tuple2<Integer, PrestoSparkRow>> rddResults;
             try {
                 rddResults = doExecute(plan);
-                commit();
+                commit(session, transactionManager);
             }
             catch (RuntimeException executionFailure) {
                 try {
-                    rollback();
+                    rollback(session, transactionManager);
                 }
                 catch (RuntimeException rollbackFailure) {
                     if (executionFailure != rollbackFailure) {
                         executionFailure.addSuppressed(rollbackFailure);
                     }
                 }
-                queryCompletedEvent(Optional.of(executionFailure));
+                try {
+                    queryCompletedEvent(Optional.of(executionFailure));
+                }
+                catch (RuntimeException eventFailure) {
+                    if (executionFailure != eventFailure) {
```

Comment:
wow. When  would this if-statement take effect? (an exception thrown during publishing the event, and the exception thrown during publishing is *same* as the exception thrown during execution).

Reply:
This is to handle some weird case when rollback throws very same exception. Exceptions cannot be self suppressed. So this is pretty standard boilerplate code.